src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs

```diff
@@ -504,40 +493,41 @@ private ReadOnlySpan<byte> InternalRead(int numBytes)
         // reasons. More about the subject in: https://github.com/dotnet/coreclr/pull/22102
         protected virtual void FillBuffer(int numBytes)
         {
-            if (numBytes < 0 || numBytes > _buffer.Length)
-            {
-                throw new ArgumentOutOfRangeException(nameof(numBytes), SR.ArgumentOutOfRange_BinaryReaderFillBuffer);
-            }
+            ArgumentOutOfRangeException.ThrowIfNegative(numBytes);
 
             ThrowIfDisposed();
 
-            // Need to find a good threshold for calling ReadByte() repeatedly
-            // vs. calling Read(byte[], int, int) for both buffered & unbuffered
-            // streams.
-            if (numBytes == 1)
-            {
-                int n = _stream.ReadByte();
-                if (n == -1)
-                {
-                    ThrowHelper.ThrowEndOfFileException();
-                }
-
-                _buffer[0] = (byte)n;
-                return;
-            }
-
-            if (numBytes > 0)
+            switch (numBytes)
             {
-                _stream.ReadExactly(_buffer.AsSpan(0, numBytes));
-            }
-            else
-            {
-                // ReadExactly no-ops for empty buffers, so special case numBytes == 0 to preserve existing behavior.
-                int n = _stream.Read(_buffer, 0, 0);
-                if (n == 0)
-                {
-                    ThrowHelper.ThrowEndOfFileException();
-                }
+                case 0:
+                    // ReadExactly no-ops for empty buffers, so special case numBytes == 0 to preserve existing behavior.
+                    int n = _stream.Read(Array.Empty<byte>(), 0, 0);
+                    if (n == 0)
+                    {
+                        ThrowHelper.ThrowEndOfFileException();
+                    }
+                    break;
+                case 1:
+                    // Need to find a good threshold for calling ReadByte() repeatedly
+                    // vs. calling Read(byte[], int, int) for both buffered & unbuffered
+                    // streams.
+                    n = _stream.ReadByte();
+                    if (n == -1)
+                    {
+                        ThrowHelper.ThrowEndOfFileException();
+                    }
+                    break;
+                default:
+                    byte[] buffer = ArrayPool<byte>.Shared.Rent(numBytes);
+                    try
+                    {
+                        _stream.ReadExactly(buffer.AsSpan(0, numBytes));
+                    }
+                    finally
+                    {
+                        ArrayPool<byte>.Shared.Return(buffer);
```

Comment:
in general I believe we don't attempt to return the buffer in exceptional cases -- saving the try/catch. Is this a case where exceptions are sufficiently unexceptional that it's worth having try/catch?

BTW I don't know whether try/catch ends up measurably impacting perf, but it presumably at least increases generated code length.